aten/src/ATen/native/cpu/Parallel.h

```diff
@@ -7,18 +8,33 @@
 namespace at {
 namespace native {
 
-constexpr size_t _THRESHOLD = 32768;
+// This parameter is heuristically chosen to determine the minimum number of
+// work that warrants paralellism. For example, when summing an array, it is
+// deemed inefficient to parallelise over arrays shorter than 32768. Further,
+// no parallel algorithm (such as parallel_reduce) should split work into
+// smaller than GRAIN_SIZE chunks.
+constexpr size_t GRAIN_SIZE = 32768;
 
 template <class T, template <class> class PRED>
 T parallel_reduce(T (*f)(const T *, size_t, size_t, T), const T *data,
                   size_t start, size_t end, T init_) {
+
+  static tbb::task_scheduler_init tbbinit;
+  static int num_threads_ = 0;
+  int num_threads = at::get_num_threads();
+  if (num_threads_ != num_threads) {
+    if (tbbinit.is_active()) tbbinit.terminate();
+    tbbinit.initialize(num_threads);
+    num_threads_ = num_threads;
+  }
```

Comment:
This is not thread safe.

Reply:
Does `task_scheduler_init` actually keep the threads alive, or is it just a small struct holding the desired config?

Reply:
It initializes the thread scheduler.

Reply:
Meaning that its creation spawns the threads? Is the scheduler global?

Reply:
At least the creation of the static local variables should be [thread safe](http://en.cppreference.com/w/cpp/language/storage_duration) since C++11. I'll guard the case when num_threads_ is not equal to num_threads.

Reply:
The most important part is that tbbinit is in scope of the parallel constructs. That's what actually sets the number of threads used for the overall [market](https://software.intel.com/en-us/blogs/2011/04/09/tbb-initialization-termination-and-resource-management-details-juicy-and-gory).